<h1 align="center">
CoolSeqTool
</h1>

<p align="center">

**[Documentation](#)** · [Installation](#) · [Usage](#) · [API reference](#)

</p>

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/cool-seq-tool?color=gr) [![tests](https://github.com/genomicmedlab/cool-seq-tool/actions/workflows/github-actions.yml/badge.svg)](https://github.com/genomicmedlab/cool-seq-tool/actions/workflows/github-actions.yml)

## Overview

<!-- description -->
The **CoolSeqTool** provides:

  - Transcript alignment data from the [UTA](https://github.com/biocommons/uta) database
  - Fast access to sequence data using [SeqRepo](https://github.com/biocommons/biocommons.seqrepo)
  - Liftover between assemblies (GRCh38 <--> GRCh37) from [PyLiftover](https://github.com/konstantint/pyliftover)
  - Lifting over to preferred [MANE](https://www.ncbi.nlm.nih.gov/refseq/MANE/) compatible transcript. See [here]() for more information.
<!-- /description -->

---

## Install

CoolSeqTool is available on [PyPI](https://pypi.org/project/cool-seq-tool)

```shell
python3 -m pip install cool-seq-tool
```

See the [installation instructions](#) in the documentation for a description of dependency setup requirements.

---

## Feedback and contributing

We welcome bug reports, feature requests, and code contributions from users and interested collaborators. The [documentation](#) contains guidance for submitting feedback and contributing new code.
